src/nixos/machines/tupac/config/disks.nix

```diff
@@ -0,0 +1,236 @@
+{ config, lib, ... }:
+
+# Nix-based Disk Management of TUPAC
+
+# Formatting strategy (Impermanence):
+#    Table: GPT
+#    2048 - 1050623 (1048576) [512M] -- EFI BOOT with FAT32
+#    1050624-1874579455 (1873528832) -- -60G Nix Store with BTRFS
+#    1874579456-2000408575 (125829120) -- Encrypted SWAP
+
+# Formatting strategy (WITHOUT impermanence):
+#    Table: GPT
+#    2048 - 1050623 (1048576) [512M] -- EFI BOOT with FAT32
+#    1050624-1874579455 (1873528832) -- -60G rootfs with BTRFS
+#    1874579456-2000408575 (125829120) -- Encrypted SWAP
+
+let
+	inherit (lib) mkMerge;
+in {
+	config = mkMerge [
+		{
+			age.secrets.tupac-disks-password.file = ../secrets/tupac-disks-password.age;
+
+			age.identityPaths = (if config.boot.impermanence.enable
+				then [ "/nix/persist/system/etc/ssh/ssh_host_ed25519_key" ]
+				else [ "/etc/ssh/ssh_host_ed25519_key" ]);
+		}
+
+		# FIXME-QA(Krey): Produces an infinite recursion -- (config.boot.impermanence.enable == true)
+		(if (true) then {
+			fileSystems."/nix/persist/system".neededForBoot = true;
+
+			# FIXME(Krey): Figure out how to do labels
+			disko.devices = {
+				nodev."/" = {
+					fsType = "tmpfs";
+					mountOptions = [
+						"size=2G"
+						"defaults"
+						# set mode to 755, otherwise systemd will set it to 777, which cause problems.
+						# relatime: Update inode access times relative to modify or change time.
+						"mode=755"
+					];
+				};
+
+				disk = {
+					system = {
+						device = "/dev/disk/by-id/nvme-SOLIDIGM_SSDPFKNU010TZ_BTEH24220RNQ1P0B"; # NVME SSD
+						type = "disk";
+						content = {
+							type = "gpt";
+							partitions = {
+
+								boot = {
+									type = "EF00"; # EFI System Partition/
+									start = "2048";
+									size = "512M";
+									priority = 1; # Needs to be first partition
+									content = {
+										type = "filesystem";
+										format = "vfat"; # FAT32
+										mountpoint = "/boot";
+									};
+								};
+
+								store = {
+									priority = 3;
+									size = "100%";
+									content = {
+										name = "nix-store";
+										type = "luks";
+										settings.allowDiscards = true;
+
+										passwordFile = config.age.secrets.tupac-disks-password.path;
+
+										initrdUnlock = true; # Add a boot.initrd.luks.devices entry for the specified disk
+
+										extraFormatArgs = [
+											"--use-random" # use true random data from /dev/random, will block until enough entropy is available
+											"--label=CRYPT_NIX"
+										];
+
+										extraOpenArgs = [
+											"--timeout 10"
+										];
+
+										content = {
+											type = "btrfs";
+											extraArgs = [ "--label NIX_STORE" ];
+											subvolumes = {
+												"@nix" = {
+													mountpoint = "/nix";
+													mountOptions = [ "compress=lzo" "noatime" ];
```

Comment:
Why not ZSTD instead? It offers better compression ratios and faster decompression speeds compared to many of those other algorithms.

Reply:
> Why not ZSTD instead? It's modern and offers better compression ratios and faster decompression speeds compared to many of those other algorithms. -- @TanvirOnGH (https://github.com/NiXium-org/NiXium/pull/74#discussion_r1660034667)

When i was testing it, zstd was less resource efficient and i couldn't find any reliable source to give me an objective reason to use something else, if you think that this could be improved then make an issue so that it can be applies infrastructure wide and i will take a look at managing that once we have https://github.com/NiXium-org/NiXium/issues/16 and https://github.com/NiXium-org/NiXium/issues/59 managed so that we can just rewrite the whole system on demand without losing functionality.